File: backend/src/main/java/com/example/backend/domain/voter/service/VoterService.java
package com.example.backend.domain.voter.service;

import java.util.List;
import java.util.stream.Collectors;

import org.springframework.stereotype.Service;

import com.example.backend.domain.group.entity.Group;
import com.example.backend.domain.group.repository.GroupRepository;
import com.example.backend.domain.groupmember.repository.GroupMemberRepository;
import com.example.backend.domain.member.entity.Member;
import com.example.backend.domain.member.repository.MemberRepository;
import com.example.backend.domain.vote.entity.Vote;
import com.example.backend.domain.vote.repository.VoteRepository;
import com.example.backend.domain.voter.dto.MemberVoteResponseDto;
import com.example.backend.domain.voter.dto.VoterDTO;
import com.example.backend.domain.voter.entity.Voter;
import com.example.backend.domain.voter.exception.VoterErrorCode;
import com.example.backend.domain.voter.exception.VoterException;
import com.example.backend.domain.voter.repository.VoterRepository;

import jakarta.transaction.Transactional;
import lombok.RequiredArgsConstructor;

@RequiredArgsConstructor
@Service
public class VoterService {
	private final VoterRepository voterRepository;
	private final VoteRepository voteRepository;
	private final MemberRepository memberRepository;
	private final GroupRepository groupRepository;
	private final GroupMemberRepository groupMemberRepository;

	// 특정 투표에 참여한 Voter 목록 조회
	public List<VoterDTO> getVotersByVoteId(Long voteId) {
		List<Voter> voters = voterRepository.findByIdVoteId(voteId);
		return voters.stream()
			.map(VoterDTO::from) // Voter 엔티티를 VoterDTO로 변환
			.collect(Collectors.toList());
	}

	// 그룹내 투표 참여 기능
	@Transactional
	public Voter addVoter(Long groupId, Long voteId, Long memberId) {
		// 투표 및 사용자 조회
		Vote vote = voteRepository.findByIdAndGroupId(voteId, groupId)
			.orElseThrow(() -> new VoterException(VoterErrorCode.NOT_FOUND_VOTE));

		Member member = memberRepository.findById(memberId)
			.orElseThrow(() -> new VoterException(VoterErrorCode.NOT_FOUND_MEMBER));

		// 사용자가 해당 모임에 속해 있는지 확인
		Group group = groupRepository.findById(groupId)
			.orElseThrow(() -> new VoterException(VoterErrorCode.NOT_FOUND_GROUP));

		boolean isMemberInGroup = groupMemberRepository.existsByGroupAndMember(group, member);
		if (!isMemberInGroup) {
			throw new VoterException(VoterErrorCode.NOT_GROUP_MEMBER);
		}

		// 중복 참여 방지
		Voter.VoterId voterId = new Voter.VoterId(memberId, voteId);
		if (voterRepository.existsById(voterId)) {
			throw new VoterException(VoterErrorCode.ALREADY_VOTED);
		}

		// Voter 엔티티 생성 및 저장
		Voter voter = new Voter(voterId, member, vote);
		return voterRepository.save(voter);
	}

	@Transactional
	public void removeVoter(Long groupId, Long voteId, Long memberId) {
		// 투표 및 사용자 조회
		Vote vote = voteRepository.findByIdAndGroupId(voteId, groupId)
			.orElseThrow(() -> new VoterException(VoterErrorCode.NOT_FOUND_VOTE));

		Member member = memberRepository.findById(memberId)
			.orElseThrow(() -> new VoterException(VoterErrorCode.NOT_FOUND_MEMBER));

		// 사용자가 해당 모임에 속해 있는지 확인
		Group group = groupRepository.findById(groupId)
			.orElseThrow(() -> new VoterException(VoterErrorCode.NOT_FOUND_GROUP));

		boolean isMemberInGroup = groupMemberRepository.existsByGroupAndMember(group, member);
		if (!isMemberInGroup) {
			throw new VoterException(VoterErrorCode.NOT_GROUP_MEMBER);
		}

		// 투표 참여 여부 확인
		Voter.VoterId voterId = new Voter.VoterId(memberId, voteId);
		if (!voterRepository.existsById(voterId)) {
			throw new VoterException(VoterErrorCode.NOT_VOTED);
		}

		// 투표 취소 처리
		voterRepository.deleteById(voterId);
	}

	public MemberVoteResponseDto getVoteIdsByGroupIdAndMemberId(Long groupId, Long memberId) {

		List<Long> voteIds = voteRepository.findAllByGroupId(groupId).stream()
			.map(Vote::getId).toList();
		List<Long> voteResultIds = voterRepository.findVoteIdsByVoteIdsAndMemberId(voteIds, memberId);

		return new MemberVoteResponseDto(voteResultIds);
	}

}

File: backend/src/main/java/com/example/backend/domain/voter/service/VoterService.kt
package com.example.backend.domain.voter.service

import com.example.backend.domain.group.repository.GroupRepository
import com.example.backend.domain.groupmember.repository.GroupMemberRepository
import com.example.backend.domain.member.repository.MemberRepository
import com.example.backend.domain.vote.repository.VoteRepository
import com.example.backend.domain.voter.dto.MemberVoteResponseDto
import com.example.backend.domain.voter.dto.VoterDTO
import com.example.backend.domain.voter.entity.Voter
import com.example.backend.domain.voter.exception.VoterErrorCode
import com.example.backend.domain.voter.exception.VoterException
import com.example.backend.domain.voter.repository.VoterRepository
import jakarta.transaction.Transactional
import org.springframework.stereotype.Service

@Service
class VoterService(
	private val voterRepository: VoterRepository,
	private val voteRepository: VoteRepository,
	private val memberRepository: MemberRepository,
	private val groupRepository: GroupRepository,
	private val groupMemberRepository: GroupMemberRepository
) {

	fun getVotersByVoteId(voteId: Long): List<VoterDTO> {
		return voterRepository.findByIdVoteId(voteId)
			.map { VoterDTO.from(it) }
	}

	@Transactional
	fun addVoter(groupId: Long, voteId: Long, memberId: Long): Voter {
		val vote = voteRepository.findByIdAndGroupId(voteId, groupId)
			.orElseThrow { VoterException(VoterErrorCode.NOT_FOUND_VOTE) }

		val member = memberRepository.findById(memberId)
			.orElseThrow { VoterException(VoterErrorCode.NOT_FOUND_MEMBER) }

		val group = groupRepository.findById(groupId)
			.orElseThrow { VoterException(VoterErrorCode.NOT_FOUND_GROUP) }

		if (!groupMemberRepository.existsByGroupAndMember(group, member)) {
			throw VoterException(VoterErrorCode.NOT_GROUP_MEMBER)
		}

		val voterId = Voter.VoterId(memberId, voteId)
		if (voterRepository.existsById(voterId)) {
			throw VoterException(VoterErrorCode.ALREADY_VOTED)
		}

		return voterRepository.save(Voter(voterId, member, vote))
	}

	@Transactional
	fun removeVoter(groupId: Long, voteId: Long, memberId: Long) {
		val vote = voteRepository.findByIdAndGroupId(voteId, groupId)
			.orElseThrow { VoterException(VoterErrorCode.NOT_FOUND_VOTE) }

		val member = memberRepository.findById(memberId)
			.orElseThrow { VoterException(VoterErrorCode.NOT_FOUND_MEMBER) }

		val group = groupRepository.findById(groupId)
			.orElseThrow { VoterException(VoterErrorCode.NOT_FOUND_GROUP) }

		if (!groupMemberRepository.existsByGroupAndMember(group, member)) {
			throw VoterException(VoterErrorCode.NOT_GROUP_MEMBER)
		}

		val voterId = Voter.VoterId(memberId, voteId)
		if (!voterRepository.existsById(voterId)) {
			throw VoterException(VoterErrorCode.NOT_VOTED)
		}

		voterRepository.deleteById(voterId)
	}

	fun getVoteIdsByGroupIdAndMemberId(groupId: Long, memberId: Long): MemberVoteResponseDto {
		val voteIds = voteRepository.findAllByGroupId(groupId).map { it.id }
		val voteResultIds = voterRepository.findVoteIdsByVoteIdsAndMemberId(voteIds, memberId)

		return MemberVoteResponseDto(voteResultIds)
	}
}
